components/places/src/sync/history/incoming_plan.rs

```diff
@@ -0,0 +1,190 @@
+        Err(e) => return IncomingPlan::Failed(record.id.clone(), e.into()),
+    }
+    // Let's get what we know about it, if anything - last 20, like desktop?
+    const MAX_RECORDS: u32 = 20;
+    let visit_tuple = match fetch_visits(conn, &url, MAX_RECORDS) {
+        Ok(v) => v,
+        Err(e) => return IncomingPlan::Failed(record.id.clone(), e.into()),
+    };
+
+    // This all seems more messy than it should be - struggling to find the
+    // correct signature for fetch_visits
+    let (existing_page, existing_visits): (Option<FetchedVisitPage>, Vec<FetchedVisit>) = match visit_tuple {
+        None => (None, Vec::new()),
+        Some((p, v)) => (Some(p), v),
+    };
+
+    let mut cur_visit_map: HashSet<(VisitTransition, Timestamp)> = HashSet::new();
+    if let Some(p) = &existing_page {
+        assert_eq!(p.guid, record.id); // XXX - we need to handle a change of guid.
+        for visit in &existing_visits {
+            // it should be impossible for us to have invalid visits locally, but...
+            let transition = match visit.visit_type {
+                Some(t) => t,
+                None => continue,
+            };
+            let date_use = clamp_visit_date(visit.visit_date);
+            cur_visit_map.insert((transition, date_use));
+        }
+    }
+    // If we already have MAX_RECORDS visits, then we will ignore incoming
+    // visits older than that. (Not really clear why we do this and why 20 is
+    // magic, but what's good enough for desktop is good enough for us at this
+    // stage.)
+    let earliest_allowed: SystemTime = if existing_visits.len() == MAX_RECORDS as usize {
+        existing_visits[existing_visits.len() - 1].visit_date.into()
+    } else {
+        UNIX_EPOCH
+    };
+
+    // work out which of the incoming records we should apply.
+    let mut to_apply = Vec::with_capacity(record.visits.len());
+    for incoming_visit in record.visits {
+        let transition = match VisitTransition::from_primitive(incoming_visit.transition) {
+            Some(v) => v,
+            None => continue,
+        };
+        let timestamp = clamp_visit_date(incoming_visit.date);
+        if earliest_allowed > timestamp.into() {
+            continue;
+        }
+        // If the entry isn't in our map we should add it.
+        let key = (transition, timestamp);
+        if !cur_visit_map.contains(&key) {
+            to_apply.push(HistoryRecordVisit { date: timestamp, transition: transition as u8 });
+            cur_visit_map.insert(key);
+        }
+    }
+    if to_apply.len() != 0 {
+        // Now we need to check the other attributes.
+        // XXX - check if we should update title? For now, assume yes.
+        let new_title = Some(record.title);
+        IncomingPlan::Apply(record.id, url.clone(), new_title, to_apply)
+    } else {
+        IncomingPlan::Skip
+    }
+}
+
+lazy_static! {
+    static ref EARLIEST_BOOKMARK_TIMESTAMP: Timestamp = Timestamp(0); // XXX - want Jan 23, 1993
```

Comment:
I think this can be `const EARLIEST_BOOKMARK_TIMESTAMP: Timestamp = Timestamp(0)` without lazy static.